src/apprunner/explorer/apprunnerNode.ts

```diff
@@ -0,0 +1,123 @@
+/*!
+ * Copyright 2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+import { makeChildrenNodes } from '../../shared/treeview/treeNodeUtilities'
+import * as vscode from 'vscode'
+import { AWSTreeNodeBase } from '../../shared/treeview/nodes/awsTreeNodeBase'
+import { AppRunnerServiceNode } from './apprunnerServiceNode'
+import { ErrorNode } from '../../shared/treeview/nodes/errorNode'
+import { PlaceholderNode } from '../../shared/treeview/nodes/placeholderNode'
+import * as nls from 'vscode-nls'
+import { AppRunnerClient } from '../../shared/clients/apprunnerClient'
+import { getPaginatedAwsCallIter } from '../../shared/utilities/collectionUtils'
+import { AppRunner } from 'aws-sdk'
+
+const localize = nls.loadMessageBundle()
+
+const POLLING_INTERVAL = 20000
+export class AppRunnerNode extends AWSTreeNodeBase {
+    private readonly serviceNodes: Map<AppRunner.ServiceId, AppRunnerServiceNode>
+    private readonly pollingNodes: Set<string> = new Set()
+    private pollTimer?: NodeJS.Timeout
+
+    public constructor(public readonly region: string, public readonly client: AppRunnerClient) {
+        super('App Runner', vscode.TreeItemCollapsibleState.Collapsed)
+        this.serviceNodes = new Map<string, AppRunnerServiceNode>()
+        this.contextValue = 'awsAppRunnerNode'
+    }
+
+    public async getChildren(): Promise<AWSTreeNodeBase[]> {
+        return await makeChildrenNodes({
+            getChildNodes: async () => {
+                await this.updateChildren()
+
+                return [...this.serviceNodes.values()]
+            },
+            getErrorNode: async (error: Error, logID: number) => new ErrorNode(this, error, logID),
+            getNoChildrenPlaceholderNode: async () =>
+                new PlaceholderNode(
+                    this,
+                    localize('AWS.explorerNode.apprunner.noServices', '[No App Runner services found]')
+                ),
+            sort: (nodeA: AppRunnerServiceNode, nodeB: AppRunnerServiceNode) =>
+                nodeA.label!.localeCompare(nodeB.label!),
+        })
+    }
+
+    private async getServiceSummaries(request: AppRunner.ListServicesRequest = {}): Promise<AppRunner.Service[]> {
+        const iterator = getPaginatedAwsCallIter({
+            awsCall: async request => await this.client.listServices(request),
+            nextTokenNames: {
+                request: 'NextToken',
+                response: 'NextToken',
+            },
+            request,
+        })
+
+        const services: AppRunner.Service[] = []
+
+        while (true) {
+            const next = await iterator.next()
+
+            next.value.ServiceSummaryList.map((summary: AppRunner.Service) => services.push(summary))
```

Comment:
Use a `forEach` instead of a `map`...`map` creates a new array.

Reply:
Yup accident. Map was being used for something else at one point.